architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,495 @@
+
+Dependency: Stage 1
+
+Exit criteria: existing tests passing. Add functional test for `rad resource create` using an existing built-in resource.
+
+#### Stage 4: Dynamic Resource Provider (Dynamic-RP)
+
+- Implement dynamic resource provider with support for CRUDL operations.
+  - No support for validation (yet).
+  - The dynamic resource provider requires the use of Recipes.
+- Implement integration tests and functional tests for Dynamic-RP's basic resource lifecycle using `rad resource create`.
+
+Dependency: Stage 2
+
+Exit criteria: new functional tests exercise the dynamic resource provider.
+
+#### Stage 5: Bicep Support
+
+- Implement a TypeSpec generator that can generate a resource provider manifest and Bicep extension manifest.
+- (If necessary): Update deployment engine logic to generate correct resource ids/URLs for user-defined types.
+- Implement functional tests that exercise user-defined types with Bicep.
+
+Dependency: Stage 4 needed for functional testing, no dependency to begin work.
+
+Exit criteria: new functional tests that exercise the dynamic resource provider with Bicep. 
+
+### Phase 2: Completing the UDT feature
+
+Note: at this point the UDT feature can do useful work end-to-end. The stages in this section make UDT feature-complete.
+
+#### Stage 6: Documentation and Samples
+
+We can begin adding documentation and samples once we reach **Stage 5**. Starting as soon as the features are useful allows us to incremental improve the documentation as more features come online.
+
+Dependency: Stage 5
+
+Exit criteria:
+
+- Tutorial created for an end-to-end of authoring a user-defined type and using it.
+- Conceptual documentation for extensibility.
+- Call-to-action for users to try it out and give feedback.
+
+#### Stage 7: Schema Validation
+
+- Accept an OpenAPI v3 schema as part of a resource type.
+- Implement OpenAPI validation in the dynamic resource provider.
+
+Dependency: Stage 4
+
+Exit criteria: new negative tests for the dynamic resource provider.
+
+#### Stage 8: Creation of community repo and contribution process
+
+At this stage we don't expect breaking changes going forward. UDTs will have reasonable guardrails in place (schema-based validation) so they are usable for many cases. 
+
+- Create a community *contrib* repository for user-defined types.
+- Create test automation to support contribution process.
+- Define maturity model.
+- Update existing `extenders` samples to use UDTs.
+- Call-to-action for users to start contributing and building the catalog of supported services + recipes.
+
+Dependency: Stage 7
+
+Exit criteria: community members and adopters try out using + contributing UDTs and give us feedback. 
+
+#### Stage 8: Safe Unregistration/Modification
+
+- Polish support for unregistering a resource provider.
```

Comment:
Here it says resource provider and below we talk about resource type. There should be some differences between the deletion/modification of a resource provider and resource type. We can talk about the different scenarios.

Reply:
Do you have a suggestion about what you'd like to see?

Reply:
The behavior for the following cases could be good to see:

- Case 1: `rad resourceprovider delete Contoso.Examples` with resources tied to it with some not in terminal state.
- Case 2: `rad resourceprovider delete Contoso.Examples` with resources tied to it with all in terminal state.
- Case 3: `rad resourcetype delete Contoso.Examples/postgreSQLDatabases` with resources tied to it with some not in terminal state.
- Case 4: `rad resourcetype delete Contoso.Examples/postgreSQLDatabases` with resources tied to it with all in terminal state.

Reply:
Thanks, we'll cover this topic when we design those features. This is too much detail for the intro document.